cmd/node-agent/main.go

```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/coreos/go-systemd/unit"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	semver "k8s.io/kubernetes/pkg/version"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	myPodNameEnvName               = "MY_POD_NAME"
+	desiredVersionAnnotation       = "updates.coreos.com/desired-version"
+	currentVersionAnnotation       = "updates.coreos.com/current-version"
```

Comment:
Circling back to this, I'm not sure I fully understand. You mention getting the kubelet version from the source (e.g. the node object we have in memory), but as you mentioned that could be the value of the self-hosted kubelet not the on-host kubelet. I'm generally in favor of deriving information rather than denormalizing it, but in this case, to me at least, it expresses explicitly just the version of the on-host kubelet, and gives insight into the cluster operator. In other words, you know when the update is completed by verifying that `current-version == desired-version`.